### ISSUE TITLE
A lot of disconnects but no reconnects

### ISSUE THREAD

**Issue by Nixellion** [miguelgrinberg/Flask-SocketIO#685]
I just got weird behaviour, after which server broke down, but that might not be the reason...

Anyway, this looks weird:

```
2018-04-07 19:55:10.029414 :: DEBUG :: dashlogger :: unknown :: Disconnect event
2018-04-07 19:55:10.032317 :: DEBUG :: dashlogger :: unknown :: m4_30902513941193/m4_30902513941193
2018-04-07 19:55:10.032526 :: DEBUG :: dashlogger :: unknown :: m4_30902513941193 None
2018-04-07 19:55:10.061720 :: DEBUG :: dashlogger :: unknown :: Disconnect event
2018-04-07 19:55:10.064122 :: DEBUG :: dashlogger :: unknown :: m4_30902513941193/m4_30902513941193
2018-04-07 19:55:10.064372 :: DEBUG :: dashlogger :: unknown :: m4_30902513941193 None
2018-04-07 19:55:10.094230 :: DEBUG :: dashlogger :: unknown :: Disconnect event
2018-04-07 19:55:10.096855 :: DEBUG :: dashlogger :: unknown :: m4_30902513941193/m4_30902513941193
2018-04-07 19:55:10.097039 :: DEBUG :: dashlogger :: unknown :: m4_30902513941193 None
2018-04-07 19:55:10.154927 :: DEBUG :: dashlogger :: unknown :: Disconnect event
2018-04-07 19:55:10.157444 :: DEBUG :: dashlogger :: unknown :: m4_30902513941193/m4_30902513941193
2018-04-07 19:55:10.157622 :: DEBUG :: dashlogger :: unknown :: m4_30902513941193 None
2018-04-07 19:55:10.342673 :: DEBUG :: dashlogger :: unknown :: Disconnect event
2018-04-07 19:55:10.345215 :: DEBUG :: dashlogger :: unknown :: m4_30902513941193/m4_30902513941193
2018-04-07 19:55:10.345395 :: DEBUG :: dashlogger :: unknown :: m4_30902513941193 None 
```

And it goes on for a few seconds. And not just for this client, but others as well. It just keeps disconnecting without connection events, and within milliseconds, I'm frustrated :D 

Granted, there was some work going on with those clients, but there's no physical way to reconnect within 0.01 ms and lose connection again...

**Comment by miguelgrinberg** [comment]
Can you add SocketIO logging to your log output?

**Comment by Nixellion** [comment]
Hello @miguelgrinberg , sorry for delay in response, had to track it down and set up logging.

The thing is, everything works perfectly fine the whole day, but every day at around 14:15-ish local time everything bugs out. Lots of disconnects coming from same or different clients is just one thing.

Another thing is that every requests to any flask page times out with nginx throwing 504 error eventually, as if main server thread was not responding. Meanwhile background threads keep working just fine.

So today server logged multiple disconnects from ALL clients all within a second:

```
2018-04-12 14:13:20,785 :: engineio :: INFO :: 54e8d80519d546db866028d4c7987c35: Received packet PING data None
2018-04-12 14:13:20,786 :: engineio :: INFO :: 54e8d80519d546db866028d4c7987c35: Sending packet PONG data None
2018-04-12 14:13:21,078 :: peewee :: DEBUG :: ('SELECT "t1"."id", "t1"."date_created", "t1"."date_updated", "t1"."date_deleted", "t1"."deleted", "t1"."user_id", "t1"."password_hash", "t1"."is_admin", "t1"."token", "t1"."telegram_chat_id", "t1"."telegram_active", "t1"."telegram_last_sent" FROM "person" AS "t1" WHERE ("t1"."user_id" = ?) LIMIT 1 OFFSET 0', ['m9_30902513949157'])
2018-04-12 14:13:21,086 :: socketio :: INFO :: emitting event "miner_disconnected" to d3ece82257874ee198262278ec780386 [/ui]
2018-04-12 14:13:21,094 :: peewee :: DEBUG :: ('SELECT "t1"."id", "t1"."date_created", "t1"."date_updated", "t1"."date_deleted", "t1"."deleted", "t1"."user_id", "t1"."password_hash", "t1"."is_admin", "t1"."token", "t1"."telegram_chat_id", "t1"."telegram_active", "t1"."telegram_last_sent" FROM "person" AS "t1" WHERE ("t1"."user_id" = ?) LIMIT 1 OFFSET 0', ['m6_30902513949201'])
2018-04-12 14:13:21,095 :: socketio :: INFO :: emitting event "miner_disconnected" to 39a128d2ca924859b9e55cc34f26abeb [/ui]
2018-04-12 14:13:21,098 :: peewee :: DEBUG :: ('SELECT "t1"."id", "t1"."date_created", "t1"."date_updated", "t1"."date_deleted", "t1"."deleted", "t1"."user_id", "t1"."password_hash", "t1"."is_admin", "t1"."token", "t1"."telegram_chat_id", "t1"."telegram_active", "t1"."telegram_last_sent" FROM "person" AS "t1" WHERE ("t1"."user_id" = ?) LIMIT 1 OFFSET 0', ['m9_30902513949157'])
2018-04-12 14:13:21,101 :: socketio :: INFO :: emitting event "miner_disconnected" to 1f135a3adac642b9a4a909c3574ec528 [/ui]
2018-04-12 14:13:21,104 :: peewee :: DEBUG :: ('SELECT "t1"."id", "t1"."date_created", "t1"."date_updated", "t1"."date_deleted", "t1"."deleted", "t1"."user_id", "t1"."password_hash", "t1"."is_admin", "t1"."token", "t1"."telegram_chat_id", "t1"."telegram_active", "t1"."telegram_last_sent" FROM "person" AS "t1" WHERE ("t1"."user_id" = ?) LIMIT 1 OFFSET 0', ['m7_30902513947576'])
2018-04-12 14:13:21,106 :: socketio :: INFO :: emitting event "miner_disconnected" to 4df148d13d8a4344a7868c5fb35a0e5e [/ui]
2018-04-12 14:13:21,109 :: peewee :: DEBUG :: ('SELECT "t1"."id", "t1"."date_created", "t1"."date_updated", "t1"."date_deleted", "t1"."deleted", "t1"."user_id", "t1"."password_hash", "t1"."is_admin", "t1"."token", "t1"."telegram_chat_id", "t1"."telegram_active", "t1"."telegram_last_sent" FROM "person" AS "t1" WHERE ("t1"."user_id" = ?) LIMIT 1 OFFSET 0', ['m3_30902513941194']) 
```

And it just goes on and on until:

```
2018-04-12 14:13:22,657 :: peewee :: DEBUG :: ('SELECT "t1"."id", "t1"."date_created", "t1"."date_updated", "t1"."date_deleted", "t1"."deleted", "t1"."user_id", "t1"."password_hash", "t1"."is_admin", "t1"."token", "t1"."telegram_chat_id", "t1"."telegram_active", "t1"."telegram_last_sent" FROM "person" AS "t1" WHERE ("t1"."user_id" = ?) LIMIT 1 OFFSET 0', ['g11_30902513948690'])
2018-04-12 14:13:22,659 :: socketio :: INFO :: emitting event "client_disconnected" to cbdd939e31e24db3882b1fc38ef2199d [/ui]
2018-04-12 14:13:22,660 :: socketio :: INFO :: emitting event "get_system_info" to all [/client]
2018-04-12 14:13:22,663 :: root :: INFO :: ############### MEMORY 24.00915392524294%| AVAILABLE (59301888) | TOTAL (246996992) ##############
2018-04-12 14:13:22,944 :: telegram.vendor.ptb_urllib3.urllib3.connectionpool :: DEBUG :: https://api.telegram.org:443 "POST /id:token/getUpdates HTTP/1.1" 200 23
2018-04-12 14:13:22,945 :: telegram.bot :: DEBUG :: No new updates found.
2018-04-12 14:13:22,945 :: telegram.bot :: DEBUG :: []
2018-04-12 14:13:22,946 :: telegram.bot :: DEBUG :: Exiting: get_updates
2018-04-12 14:13:22,946 :: telegram.bot :: DEBUG :: Entering: get_updates
2018-04-12 14:13:32,664 :: socketio :: INFO :: emitting event "get_system_info" to all [/client]
2018-04-12 14:13:32,667 :: root :: INFO :: ############### MEMORY 24.28111837086664%| AVAILABLE (59973632) | TOTAL (246996992) ##############
2018-04-12 14:13:32,994 :: telegram.vendor.ptb_urllib3.urllib3.connectionpool :: DEBUG :: https://api.telegram.org:443 "POST /id:token/getUpdates HTTP/1.1" 200 23
2018-04-12 14:13:32,995 :: telegram.bot :: DEBUG :: No new updates found.
2018-04-12 14:13:32,995 :: telegram.bot :: DEBUG :: []
2018-04-12 14:13:32,996 :: telegram.bot :: DEBUG :: Exiting: get_updates
2018-04-12 14:13:32,996 :: telegram.bot :: DEBUG :: Entering: get_updates
2018-04-12 14:13:32,997 :: telegram.vendor.ptb_urllib3.urllib3.connectionpool :: DEBUG :: Resetting dropped connection: api.telegram.org
2018-04-12 14:13:42,669 :: socketio :: INFO :: emitting event "get_system_info" to all [/client] 
```

So it just goes back to it's background task duties, it emits events, and after 2 minutes starts sending telegram notifications that it did not receive any updates from previously connected clients for a while.


Prior to all that it was just working as usual. Broadcasting 'get_system_info' every couple minutes, then receiving responses from all clients. 

Clients use Python Socketio client library, and set to reconnect each hour. But it can't be that they all do that at once, as they are all launched at different times, manually.

**Comment by miguelgrinberg** [comment]
Can you capture another of these sessions also adding `engineio_logger=True` to your `SocketIO` class? That will log lower level connection stuff,

**Comment by Nixellion** [comment]
ok, i'll try

**Comment by Nixellion** [comment]
Not sure if I can see much more new entries now:

```
2018-04-13 12:38:05,992 :: engineio :: INFO :: bdf3b679ec564d0ead00b9f20561548f: Sending packet PONG data None
2018-04-13 12:38:06,040 :: engineio :: INFO :: 43a997b453fb4da7a0b542ba0fc1345e: Received packet PING data None
2018-04-13 12:38:06,040 :: engineio :: INFO :: 43a997b453fb4da7a0b542ba0fc1345e: Sending packet PONG data None
2018-04-13 12:38:06,042 :: engineio :: INFO :: ee2d5491c5124898a6ac13e639508566: Sending packet OPEN data {'upgrades': ['websocket'], 'pingTimeout': 60000, 'pingInterval': 25000, 'sid': 'ee2d5491c5124898a6ac13e639508566'}
2018-04-13 12:38:06,042 :: engineio :: INFO :: ee2d5491c5124898a6ac13e639508566: Sending packet MESSAGE data 0
2018-04-13 12:38:06,050 :: engineio :: INFO :: ee2d5491c5124898a6ac13e639508566: Received request to upgrade to websocket
2018-04-13 12:38:06,054 :: engineio :: INFO :: ee2d5491c5124898a6ac13e639508566: Sending packet NOOP data None
2018-04-13 12:38:06,056 :: engineio :: INFO :: ee2d5491c5124898a6ac13e639508566: Upgrade to websocket successful
2018-04-13 12:38:06,057 :: engineio :: INFO :: ee2d5491c5124898a6ac13e639508566: Received packet PING data None
2018-04-13 12:38:06,057 :: engineio :: INFO :: ee2d5491c5124898a6ac13e639508566: Sending packet PONG data None
2018-04-13 12:38:06,058 :: engineio :: INFO :: ee2d5491c5124898a6ac13e639508566: Received packet MESSAGE data 0/client,
2018-04-13 12:38:06,058 :: engineio :: INFO :: ee2d5491c5124898a6ac13e639508566: Sending packet MESSAGE data 0/client
2018-04-13 12:38:06,058 :: engineio :: INFO :: ee2d5491c5124898a6ac13e639508566: Received packet PING data None
2018-04-13 12:38:06,059 :: engineio :: INFO :: ee2d5491c5124898a6ac13e639508566: Sending packet PONG data None
2018-04-13 12:38:06,277 :: peewee :: DEBUG :: ('SELECT "t1"."id", "t1"."date_created", "t1"."date_updated", "t1"."date_deleted", "t1"."deleted", "t1"."user_id", "t1"."password_hash", "t1"."is_admin", "t1"."token", "t1"."telegram_chat_id", "t1"."telegram_active", "t1"."telegram_last_sent" FROM "person" AS "t1" WHERE ("t1"."user_id" = ?) LIMIT 1 OFFSET 0', ['g6_30902513948729']) 
2018-04-13 12:38:06,295 :: socketio :: INFO :: emitting event "client_disconnected" to 365b41f7469b4d8a89a9ce23692a44e0 [/ui] 

```
Then there's flood of same two messages for all disconnects, repeating for same clients over and over until it just goes back to background tasks, like before. Emitting socketio events and contacting telegram, writing logs, but not receiving anything, and website does not open. I'm not sure if engineio logging was properly activated though, I used this:

`socketio = SocketIO(app, async_mode='eventlet', engineio_logger=True)`

is this correct? I only see INFO level from engineio. 


Here's my logging setup:

```
import logging
from logging.handlers import RotatingFileHandler

logger = logging.getLogger()
logger.setLevel(logging.DEBUG)

fileHandler = RotatingFileHandler(os.path.join(realpath, 'logs', 'full_log.log'), mode='a+', maxBytes = 10*1024*1024, backupCount=2, encoding='utf-8', delay=0)

formatter = logging.Formatter('%(asctime)s :: %(name)s :: %(levelname)s :: %(message)s')
fileHandler.setFormatter(formatter)

logger.addHandler(fileHandler)
logger.addHandler(logging.StreamHandler())
```

Here's log from server restart:

```
2018-04-13 15:50:01,966 :: engineio :: INFO :: Server initialized for eventlet.
2018-04-13 15:50:03,176 :: telegram.ext.dispatcher :: DEBUG :: Setting singleton dispatcher as
2018-04-13 15:50:03,454 :: socketio :: INFO :: emitting event "get_system_info" to all [/client]
2018-04-13 15:50:03,458 :: root :: INFO :: ############### MEMORY 54.596862458956586%| AVAILABLE (134852608) | TOTAL (246996992) ##############
2018-04-13 15:50:03,470 :: JobQueue :: DEBUG :: JobQueue thread started
2018-04-13 15:50:03,471 :: telegram.ext.updater :: DEBUG :: MainThread - started
2018-04-13 15:50:03,472 :: telegram.ext.updater :: DEBUG :: MainThread - started
2018-04-13 15:50:03,472 :: telegram.ext.updater :: DEBUG :: Updater thread started (polling)
2018-04-13 15:50:03,473 :: telegram.ext.updater :: DEBUG :: Start network loop retry bootstrap del webhook
2018-04-13 15:50:03,473 :: telegram.bot :: DEBUG :: Entering: delete_webhook
2018-04-13 15:50:03,474 :: telegram.vendor.ptb_urllib3.urllib3.connectionpool :: DEBUG :: Starting new HTTPS connection (1): api.telegram.org
2018-04-13 15:50:03,476 :: telegram.ext.dispatcher :: DEBUG :: Dispatcher started
2018-04-13 15:50:03,669 :: telegram.vendor.ptb_urllib3.urllib3.connectionpool :: DEBUG :: https://api.telegram.org:443 "POST /bot573819906:AAF_rMCCBWIHP83eTzNqKGgIUghVDc-Nda4/deleteWebhook HTTP/1.1" 200 68
2018-04-13 15:50:03,670 :: telegram.bot :: DEBUG :: True
2018-04-13 15:50:03,671 :: telegram.bot :: DEBUG :: Exiting: delete_webhook
2018-04-13 15:50:03,671 :: telegram.ext.updater :: DEBUG :: Bootstrap done
2018-04-13 15:50:03,672 :: telegram.ext.updater :: DEBUG :: Start network loop retry getting Updates
2018-04-13 15:50:03,672 :: telegram.bot :: DEBUG :: Entering: get_updates
2018-04-13 15:50:03,884 :: engineio :: INFO :: 764859eae0a1465bb70b9431b554e7d7: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '764859eae0a1465bb70b9431b554e7d7', 'pingInterval': 25000}
2018-04-13 15:50:03,885 :: engineio :: INFO :: 764859eae0a1465bb70b9431b554e7d7: Sending packet MESSAGE data 0
2018-04-13 15:50:03,887 :: engineio :: INFO :: b973fb0e8cea4d139c9199b08b4872cb: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': 'b973fb0e8cea4d139c9199b08b4872cb', 'pingInterval': 25000}
2018-04-13 15:50:03,888 :: engineio :: INFO :: b973fb0e8cea4d139c9199b08b4872cb: Sending packet MESSAGE data 0
2018-04-13 15:50:03,889 :: engineio :: INFO :: a4b33d8715914cdc8146f1def8ebc361: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': 'a4b33d8715914cdc8146f1def8ebc361', 'pingInterval': 25000}
2018-04-13 15:50:03,890 :: engineio :: INFO :: a4b33d8715914cdc8146f1def8ebc361: Sending packet MESSAGE data 0
2018-04-13 15:50:03,892 :: engineio :: INFO :: 59f0ec2b3a7e4b4b86733d3c447a9199: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '59f0ec2b3a7e4b4b86733d3c447a9199', 'pingInterval': 25000}
2018-04-13 15:50:03,893 :: engineio :: INFO :: 59f0ec2b3a7e4b4b86733d3c447a9199: Sending packet MESSAGE data 0
2018-04-13 15:50:03,895 :: engineio :: INFO :: 32a74b2cceb94a12ac9ca3be62ebc1cb: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '32a74b2cceb94a12ac9ca3be62ebc1cb', 'pingInterval': 25000}
2018-04-13 15:50:03,897 :: engineio :: INFO :: 32a74b2cceb94a12ac9ca3be62ebc1cb: Sending packet MESSAGE data 0
2018-04-13 15:50:03,898 :: engineio :: INFO :: 5579e487e8e64f91a47b091eba78c8c1: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '5579e487e8e64f91a47b091eba78c8c1', 'pingInterval': 25000}
2018-04-13 15:50:03,899 :: engineio :: INFO :: 5579e487e8e64f91a47b091eba78c8c1: Sending packet MESSAGE data 0
2018-04-13 15:50:03,900 :: engineio :: INFO :: 764859eae0a1465bb70b9431b554e7d7: Received request to upgrade to websocket
2018-04-13 15:50:03,901 :: engineio :: INFO :: ddeba2a491c149b19a22d64669d7215c: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': 'ddeba2a491c149b19a22d64669d7215c', 'pingInterval': 25000}
2018-04-13 15:50:03,902 :: engineio :: INFO :: ddeba2a491c149b19a22d64669d7215c: Sending packet MESSAGE data 0
2018-04-13 15:50:03,904 :: engineio :: INFO :: b973fb0e8cea4d139c9199b08b4872cb: Received request to upgrade to websocket
2018-04-13 15:50:03,905 :: engineio :: INFO :: 6756503bb2784826a5b43a4ed5d49058: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '6756503bb2784826a5b43a4ed5d49058', 'pingInterval': 25000}
2018-04-13 15:50:03,906 :: engineio :: INFO :: 6756503bb2784826a5b43a4ed5d49058: Sending packet MESSAGE data 0
2018-04-13 15:50:03,907 :: engineio :: INFO :: 764859eae0a1465bb70b9431b554e7d7: Sending packet NOOP data None
2018-04-13 15:50:03,909 :: engineio :: INFO :: acf1320bc90b4051a8fe54c15eb3bd50: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': 'acf1320bc90b4051a8fe54c15eb3bd50', 'pingInterval': 25000}
2018-04-13 15:50:03,912 :: engineio :: INFO :: acf1320bc90b4051a8fe54c15eb3bd50: Sending packet MESSAGE data 0
2018-04-13 15:50:03,914 :: engineio :: INFO :: 15e7cf6930b0409ea3bb96c5a2d01153: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '15e7cf6930b0409ea3bb96c5a2d01153', 'pingInterval': 25000}
2018-04-13 15:50:03,915 :: engineio :: INFO :: 15e7cf6930b0409ea3bb96c5a2d01153: Sending packet MESSAGE data 0
2018-04-13 15:50:03,917 :: engineio :: INFO :: 59f0ec2b3a7e4b4b86733d3c447a9199: Received request to upgrade to websocket
2018-04-13 15:50:03,918 :: engineio :: INFO :: a4b33d8715914cdc8146f1def8ebc361: Received request to upgrade to websocket
2018-04-13 15:50:03,919 :: engineio :: INFO :: d080b7f4fdd14d64a47b253a8577da41: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': 'd080b7f4fdd14d64a47b253a8577da41', 'pingInterval': 25000}
2018-04-13 15:50:03,920 :: engineio :: INFO :: d080b7f4fdd14d64a47b253a8577da41: Sending packet MESSAGE data 0
2018-04-13 15:50:03,921 :: engineio :: INFO :: ea9b5eae0bd94acc9b5797269c4d82a4: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': 'ea9b5eae0bd94acc9b5797269c4d82a4', 'pingInterval': 25000}
2018-04-13 15:50:03,922 :: engineio :: INFO :: ea9b5eae0bd94acc9b5797269c4d82a4: Sending packet MESSAGE data 0
2018-04-13 15:50:03,923 :: engineio :: INFO :: a07bf68866fe4b7991c18ba582b7b2d0: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': 'a07bf68866fe4b7991c18ba582b7b2d0', 'pingInterval': 25000}
2018-04-13 15:50:03,924 :: engineio :: INFO :: a07bf68866fe4b7991c18ba582b7b2d0: Sending packet MESSAGE data 0
2018-04-13 15:50:03,925 :: engineio :: INFO :: fe4ab0cd0f7f4d95baa3d3d34ec5f8c8: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': 'fe4ab0cd0f7f4d95baa3d3d34ec5f8c8', 'pingInterval': 25000}
2018-04-13 15:50:03,926 :: engineio :: INFO :: fe4ab0cd0f7f4d95baa3d3d34ec5f8c8: Sending packet MESSAGE data 0
2018-04-13 15:50:03,927 :: engineio :: INFO :: 03fe8fed8b904eca85926f1f5523aaed: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '03fe8fed8b904eca85926f1f5523aaed', 'pingInterval': 25000}
2018-04-13 15:50:03,928 :: engineio :: INFO :: 03fe8fed8b904eca85926f1f5523aaed: Sending packet MESSAGE data 0
2018-04-13 15:50:03,929 :: engineio :: INFO :: 32a74b2cceb94a12ac9ca3be62ebc1cb: Received request to upgrade to websocket
2018-04-13 15:50:03,931 :: engineio :: INFO :: 3cda078d5222441faf14a2a7f32b8ce0: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '3cda078d5222441faf14a2a7f32b8ce0', 'pingInterval': 25000}
2018-04-13 15:50:03,931 :: engineio :: INFO :: 3cda078d5222441faf14a2a7f32b8ce0: Sending packet MESSAGE data 0
2018-04-13 15:50:03,932 :: engineio :: INFO :: 5579e487e8e64f91a47b091eba78c8c1: Received request to upgrade to websocket
2018-04-13 15:50:03,935 :: engineio :: INFO :: 32a74b2cceb94a12ac9ca3be62ebc1cb: Sending packet NOOP data None
2018-04-13 15:50:03,940 :: engineio :: INFO :: 59f0ec2b3a7e4b4b86733d3c447a9199: Sending packet NOOP data None
2018-04-13 15:50:03,940 :: engineio :: INFO :: a4b33d8715914cdc8146f1def8ebc361: Sending packet NOOP data None
2018-04-13 15:50:03,940 :: engineio :: INFO :: b973fb0e8cea4d139c9199b08b4872cb: Sending packet NOOP data None
2018-04-13 15:50:03,941 :: engineio :: INFO :: 45c91533e77841fa8a327f81113ba0bd: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '45c91533e77841fa8a327f81113ba0bd', 'pingInterval': 25000}
2018-04-13 15:50:03,941 :: engineio :: INFO :: 45c91533e77841fa8a327f81113ba0bd: Sending packet MESSAGE data 0
2018-04-13 15:50:03,947 :: engineio :: INFO :: 08cda156d9224f6caba0e181caab2904: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '08cda156d9224f6caba0e181caab2904', 'pingInterval': 25000}
2018-04-13 15:50:03,948 :: engineio :: INFO :: 08cda156d9224f6caba0e181caab2904: Sending packet MESSAGE data 0
2018-04-13 15:50:03,950 :: engineio :: INFO :: d34a7127f990499583049d3f74c28ee2: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': 'd34a7127f990499583049d3f74c28ee2', 'pingInterval': 25000}
2018-04-13 15:50:03,950 :: engineio :: INFO :: d34a7127f990499583049d3f74c28ee2: Sending packet MESSAGE data 0
2018-04-13 15:50:03,952 :: engineio :: INFO :: ddeba2a491c149b19a22d64669d7215c: Received request to upgrade to websocket
2018-04-13 15:50:03,954 :: engineio :: INFO :: 4e6244aeb51f451c9bdcc6932b06426a: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '4e6244aeb51f451c9bdcc6932b06426a', 'pingInterval': 25000}
2018-04-13 15:50:03,954 :: engineio :: INFO :: 4e6244aeb51f451c9bdcc6932b06426a: Sending packet MESSAGE data 0
2018-04-13 15:50:03,956 :: engineio :: INFO :: 6756503bb2784826a5b43a4ed5d49058: Received request to upgrade to websocket
2018-04-13 15:50:03,957 :: engineio :: INFO :: 5f4effe7b4e141a095e4524d1dcb7abc: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '5f4effe7b4e141a095e4524d1dcb7abc', 'pingInterval': 25000}
2018-04-13 15:50:03,958 :: engineio :: INFO :: 5f4effe7b4e141a095e4524d1dcb7abc: Sending packet MESSAGE data 0
2018-04-13 15:50:03,959 :: engineio :: INFO :: acf1320bc90b4051a8fe54c15eb3bd50: Received request to upgrade to websocket
2018-04-13 15:50:03,960 :: engineio :: INFO :: 2d76c624c6f647afb398dbbfde8c8353: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '2d76c624c6f647afb398dbbfde8c8353', 'pingInterval': 25000}
2018-04-13 15:50:03,961 :: engineio :: INFO :: 2d76c624c6f647afb398dbbfde8c8353: Sending packet MESSAGE data 0
2018-04-13 15:50:03,962 :: engineio :: INFO :: f1321e70547c4d17b709cc639f9bb48d: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': 'f1321e70547c4d17b709cc639f9bb48d', 'pingInterval': 25000}
2018-04-13 15:50:03,963 :: engineio :: INFO :: f1321e70547c4d17b709cc639f9bb48d: Sending packet MESSAGE data 0
2018-04-13 15:50:03,964 :: engineio :: INFO :: 15e7cf6930b0409ea3bb96c5a2d01153: Received request to upgrade to websocket
2018-04-13 15:50:03,965 :: engineio :: INFO :: ea9b5eae0bd94acc9b5797269c4d82a4: Received request to upgrade to websocket
2018-04-13 15:50:03,966 :: engineio :: INFO :: 130ec4ffa61540d5b46989349cc05693: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '130ec4ffa61540d5b46989349cc05693', 'pingInterval': 25000}
2018-04-13 15:50:03,967 :: engineio :: INFO :: 130ec4ffa61540d5b46989349cc05693: Sending packet MESSAGE data 0
2018-04-13 15:50:03,968 :: engineio :: INFO :: d080b7f4fdd14d64a47b253a8577da41: Received request to upgrade to websocket
2018-04-13 15:50:03,969 :: engineio :: INFO :: a07bf68866fe4b7991c18ba582b7b2d0: Received request to upgrade to websocket
2018-04-13 15:50:03,971 :: engineio :: INFO :: 2ee2c79e7eaa488f889f3ddf084ee567: Sending packet OPEN data {'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '2ee2c79e7eaa488f889f3ddf084ee567', 'pingInterval': 25000}
2018-04-13 15:50:03,971 :: engineio :: INFO :: 2ee2c79e7eaa488f889f3ddf084ee567: Sending packet MESSAGE data 0 
```


It repeats for all clients, then they send login events, and well, everything goes as expected.


Could it be a low-memory issue or something like that? I'm logging memory percentage every 10-15 seconds, and it rarely goes above 50%, but it's only 256mbs of ram. Yesterday I increased swaparea to 1256mb just in case, did not help.

What's weird is that it happens like if by a clock, so far always between 12:00 and 14:00... ish. And it does not seem to be uptime issue. Once I only did a restart at 15:00, and it happened around 12:00-14:00, another time i did a restart at 03:00 and still it failed at around 12:00-14:00.

**Comment by Nixellion** [comment]
Oh, wait. Actually even worse, I've setup a cron job to restart service twice a day, at 0 and at 12 hours, so it was only up for 1-2 hours, and still failed at 12-14

**Comment by miguelgrinberg** [comment]
Can you show me the code around that emits the `client_disconnected` event? I was hoping that the logs would show justification for the disconnect, but there is nothing.

**Comment by Nixellion** [comment]
Sure, but there's not much in there:

```
@socketio.on('disconnect', namespace='/client')
@catch_errors
def socket_disconnect(data=None):
    log.debug("Disconnect event")
    log.debug(current_user)
    log.debug(current_user.id, current_user.token)
    print("client disconnected:", current_user.id)
    client_connections.pop(request.sid, None)
    emit("client_disconnected", {'client_id': current_user.id}, namespace="/ui", room=current_user.token)
```

**Comment by Nixellion** [comment]
I think I should repeat, that the actual problem now is not in disconnect events, I think it's the **cause**.

**The problem is** that main server thread stops processing any incoming connections, so clients can't connect to it, and web pages won't open. This 'hanging' of the main thread causes clients to start disconnecting. I can open another issue with a better matching title, or rename this one, if it's necessary.

Providing full log session is not currently possible, as such log file will exceed the amount of space I have on the server. But there's literally nothing else that stands out in the logs. I only cut out repeating parts of the log. I provided logs from the moment of restart of the server, then it loops and at some moment it just stops receiving any flask-socketio AND flask (app.route) connections from outside.

Log output matches disconnect handler. I use two separate loggers, one 'log' (which logs events only relevant to my code) and another 'logger' (which records everything else). Here's disconnect event in another log file:

```
2018-04-12 14:13:21.580591 :: DEBUG :: dashlogger :: unknown :: Disconnect event
2018-04-12 14:13:21.583850 :: DEBUG :: dashlogger :: unknown :: g7_30902513948557/g7_30902513948557
2018-04-12 14:13:21.584027 :: DEBUG :: dashlogger :: unknown :: g7_30902513948557 None 
```

**Comment by miguelgrinberg** [comment]
Sorry, completely missed the second part, where clients do not reconnect. Have you looked at one of the clients to see what happens when they try to reconnect? Are they sending the connection request?

**Comment by Nixellion** [comment]
Yes, I'm certain that they keep trying to reconnect, but fail. Can't check client logs at the moment, and clients are on Python Socketio.

But then again, even @app.route("/") does not respond (can't open it in web browser), and eventually times out with nginx 504 error. Until I restart the service.

Other things on the server keep working fine: ssh, webmin, flask background tasks, etc

**Comment by miguelgrinberg** [comment]
Could you be reaching some maximum on open file handlers? Is it related to number of concurrent clients in any way?

**Comment by Nixellion** [comment]
I do use my own logger which is using  `with open() as f:` to write to a file, and it's the only approach to openning files I generally use, so they should close once write is finished. I think it's the only thing that's constantly writing\reading.

Another is peewee, with sqlite.

I don't think it's related to number of concurrent clients, there's about 25 clients all the time. Rarely more than 30, at least now.

**Comment by miguelgrinberg** [comment]
File handlers include network sockets, but yeah, 25-30 clients should not get you into those limits.

You are using eventlet, so I guess it is possible that something in your app blocks. That would make the entire process stop processing requests, but it's really hard to know what happened. Maybe if you look at the stack trace after a Ctrl-C you can find what the process is doing.

**Comment by Nixellion** [comment]
Yes, I'm certain that they keep trying to reconnect, but fail. Can't check client logs at the moment, and clients are on Python Socketio.

But then again, even @app.route("/") does not respond (can't open it in web browser), and eventually times out with nginx 504 error. Until I restart the service.

Other things on the server keep working fine: ssh, webmin, flask background tasks, etc

**Comment by Nixellion** [comment]
Yes, I think I also at some point thought about it too, as it seems, because here's current limit:

`open files                      (-n) 1048576`

Hm, the way you say it about eventlet made me wonder - is there any other non-blocking option? Threading? Or multiple workers I suppose? Of course it's not a solution to the problem, just a thought.

I'm not sure about Ctrl-C... You mean - launch the process manually? It's currently running as a service. I did look at `systemctl status `output after each crash, nothing special there either.

I'm running telegram bot as a background process, could it be that it blocks? Do I need to run monkey patch in each module or only the `app.py` (main) one?

**Comment by miguelgrinberg** [comment]
> I'm not sure about Ctrl-C... You mean - launch the process manually?

Yes, start it in the foreground in your console, wait until clients start to hang, then hit Ctrl-C to see if you can get a stack trace that points to the blockage.

> I'm running telegram bot as a background process, could it be that it blocks?

This is a separate process, right? You should look at your server, other processes blocking would not immediately cause your server to block.

> Do I need to run monkey patch in each module or only the app.py (main) one?

You need to monkey patch in your main script. That applies to the entire process. If you are using eventlet in multiple process, then monkey patch each one of them when they start.

**Comment by Nixellion** [comment]
> Yes, start it in the foreground in your console, wait until clients start to hang, then hit Ctrl-C to see if you can get a stack trace that points to the blockage.

Well might have to leave it on for 24 hours, as unless it happens on it's usual schedule. I'll try

> This is a separate process, right? You should look at your server, other processes blocking would not immediately cause your server to block.

I run it with `socketio.start_background_task()`

**Comment by miguelgrinberg** [comment]
Okay, so background tasks can definitely block your server, because they run in the same process. The monkey patching that you do at the start does apply to these, but monkey patching does not solve all blocking problems.

**Comment by Nixellion** [comment]
Well, I turned off background tasks, and it still happens.

I actually think that assumption that main thread blocks might be not right, because if it did, then I would not get disconnects events logged by server, right? I mean, if main thread is stuck, it would not be able to log disconnect events to a file, until it stops blocking?

And then, if it stopped blocking clients would be able to reconnect. But it does not happen. 

I'll keep trying to find clues and issue, thanks for the help so far!

**Comment by miguelgrinberg** [comment]
I guess it really depends on how the application blocks, if it blocks at all. I agree that the most common result is for the application to stop responding to anything, including logging.

**Comment by Nixellion** [comment]
Well, new catch:

```
2018-04-18 03:06:44,655 :: engineio :: ERROR :: message handler error
Traceback (most recent call last):
File "/opt/midas_dashboard/app.py", line 291, in wrapped
File "/opt/midas_dashboard/app.py", line 373, in login_fail
OSError: [Errno 24] Too many open files: '/opt/midas_dashboard/visitors.log'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
File "/opt/midas_dashboard/app.py", line 307, in wrapped
File "/opt/midas_dashboard/app.py", line 298, in wrapped
File "/usr/local/lib/python3.5/dist-packages/flask/templating.py", line 133, in render_template
File "/usr/local/lib/python3.5/dist-packages/jinja2/environment.py", line 869, in get_or_select_template
File "/usr/local/lib/python3.5/dist-packages/jinja2/environment.py", line 830, in get_template
File "/usr/local/lib/python3.5/dist-packages/jinja2/environment.py", line 804, in _load_template
File "/usr/local/lib/python3.5/dist-packages/jinja2/loaders.py", line 113, in load
File "/usr/local/lib/python3.5/dist-packages/flask/templating.py", line 57, in get_source
File "/usr/local/lib/python3.5/dist-packages/flask/templating.py", line 82, in _get_source_fast
File "/usr/local/lib/python3.5/dist-packages/jinja2/loaders.py", line 171, in get_source
File "/usr/local/lib/python3.5/dist-packages/jinja2/utils.py", line 154, in open_if_exists
OSError: [Errno 24] Too many open files: '/opt/midas_dashboard/templates/pages/500.html'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
File "/usr/local/lib/python3.5/dist-packages/engineio/server.py", line 405, in _trigger_event
File "/usr/local/lib/python3.5/dist-packages/socketio/server.py", line 522, in _handle_eio_message
File "/usr/local/lib/python3.5/dist-packages/socketio/server.py", line 458, in _handle_event
File "/usr/local/lib/python3.5/dist-packages/socketio/server.py", line 461, in _handle_event_internal
File "/usr/local/lib/python3.5/dist-packages/socketio/server.py", line 490, in _trigger_event
File "/usr/local/lib/python3.5/dist-packages/flask_socketio/__init__.py", line 243, in _handler
File "/usr/local/lib/python3.5/dist-packages/flask_socketio/__init__.py", line 626, in _handle_event
File "/opt/midas_dashboard/app.py", line 310, in wrapped
File "/opt/midas_dashboard/debug.py", line 75, in error
File "/opt/midas_dashboard/debug.py", line 58, in log
OSError: [Errno 24] Too many open files: '/opt/midas_dashboard/logs/dash_log.log' 
```

**Comment by miguelgrinberg** [comment]
> OSError: [Errno 24] Too many open files: '/opt/midas_dashboard/templates/pages/500.html'

You have a file handle leak in your application is my guess.

**Comment by Nixellion** [comment]
Yes, I suppose so. I limited the use of 'login_fail ' for now, I think I was calling it too often in some cases. But if it's not the problem it might just delay the problem.

`with open` is supposed to close the file, right? I never use `f = open(); f.close()`, so that can't be a leak.

Will keep looking for the root of the problem.

**Comment by miguelgrinberg** [comment]
Yes, the `with open()` way of working with files is safe against leaks. You may find some information by inspecting the list of open file handlers with `lsof` or similar tool.

**Comment by miguelgrinberg** [comment]
So you need to watch the list of open handler over time and see if it is growing or if it is in a steady state. If it grows, see which kind of handlers are growing. That should give you an idea of where the potential leak might be.

**Comment by Nixellion** [comment]
I noticed that process had ulimit of 1024, because it was not using default limits. After adding `LimitNOFILE=1024000` to the *.service file it now uses this number.

Will keep an eye on number of open handles.

**Comment by Nixellion** [comment]
Well, so far I only see that TCP connections and database entries grow. It's been 144, after about 5-6 hours it's 400+ open handles. At least when using `lsof -p` command. Not sure if this the correct way to check it.

This is the part that crows (copy paste over and over to get the idea :D )

```
python3 1294 root  115u     IPv4              18700      0t0    TCP midash.net:14864->212.118.33.25:1635 (ESTABLISHED)
python3 1294 root  116u     IPv4              18702      0t0    TCP midash.net:14864->212.118.33.25:55037 (ESTABLISHED)
python3 1294 root  117u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  118u     IPv4              18709      0t0    TCP midash.net:14864->212.118.33.25:56260 (ESTABLISHED)
python3 1294 root  119u     IPv4              18707      0t0    TCP midash.net:14864->212.118.33.25:56259 (ESTABLISHED)
python3 1294 root  120u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  121u     IPv4              18713      0t0    TCP midash.net:14864->212.118.33.25:55113 (ESTABLISHED)
python3 1294 root  122u     IPv4              18715      0t0    TCP midash.net:14864->212.118.33.25:55114 (ESTABLISHED)
python3 1294 root  123u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  124u     IPv4              18719      0t0    TCP midash.net:14864->212.118.33.25:55158 (ESTABLISHED)
python3 1294 root  125u     IPv4              18721      0t0    TCP midash.net:14864->212.118.33.25:55159 (ESTABLISHED)
python3 1294 root  126u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  127u     IPv4              18725      0t0    TCP midash.net:14864->212.118.33.25:55169 (ESTABLISHED)
python3 1294 root  128u     IPv4              18727      0t0    TCP midash.net:14864->212.118.33.25:55170 (ESTABLISHED)
python3 1294 root  129u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  130u     IPv4              18733      0t0    TCP midash.net:14864->212.118.33.25:1637 (ESTABLISHED)
python3 1294 root  131u     IPv4              18731      0t0    TCP midash.net:14864->212.118.33.25:1636 (ESTABLISHED)
python3 1294 root  132u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  133u     IPv4              18737      0t0    TCP midash.net:14864->212.118.33.25:55184 (ESTABLISHED)
python3 1294 root  134u     IPv4              18739      0t0    TCP midash.net:14864->212.118.33.25:55185 (ESTABLISHED)
python3 1294 root  135u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  136u     IPv4              18745      0t0    TCP midash.net:14864->212.118.33.25:55181 (ESTABLISHED)
python3 1294 root  137u     IPv4              18747      0t0    TCP midash.net:14864->212.118.33.25:55182 (ESTABLISHED)
python3 1294 root  138u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  139u     IPv4              18751      0t0    TCP midash.net:14864->212.118.33.25:54464 (ESTABLISHED)
python3 1294 root  140u     IPv4              18753      0t0    TCP midash.net:14864->212.118.33.25:54465 (ESTABLISHED)
python3 1294 root  141u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  142u     IPv4              18757      0t0    TCP midash.net:14864->212.118.33.25:54557 (ESTABLISHED)
python3 1294 root  143u     IPv4              18759      0t0    TCP midash.net:14864->212.118.33.25:54558 (ESTABLISHED)
python3 1294 root  144u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  145u     IPv4              18764      0t0    TCP midash.net:14864->212.118.33.25:54604 (ESTABLISHED)
python3 1294 root  146u     IPv4              18766      0t0    TCP midash.net:14864->212.118.33.25:54605 (ESTABLISHED)
python3 1294 root  147u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  148u     IPv4              18770      0t0    TCP midash.net:14864->212.118.33.25:54523 (ESTABLISHED)
python3 1294 root  149u     IPv4              18772      0t0    TCP midash.net:14864->212.118.33.25:54524 (ESTABLISHED)
python3 1294 root  150u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  151u     IPv4              18776      0t0    TCP midash.net:14864->212.118.33.25:54448 (ESTABLISHED)
python3 1294 root  152u     IPv4              18778      0t0    TCP midash.net:14864->212.118.33.25:54449 (ESTABLISHED)
python3 1294 root  153u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  154u     IPv4              18782      0t0    TCP midash.net:14864->212.118.33.25:58715 (ESTABLISHED)
python3 1294 root  155u     IPv4              18784      0t0    TCP midash.net:14864->212.118.33.25:58716 (ESTABLISHED)
python3 1294 root  156u      REG              202,1    12288 314508 /opt/midas_dashboard/users.db
python3 1294 root  157u     IPv4              18790      0t0    TCP midash.net:14864->212.118.33.25:54434 (ESTABLISHED)
python3 1294 root  158u     IPv4              18792      0t0    TCP midash.net:14864->212.118.33.25:54435 (ESTABLISHED)

```

I suppose that it's the same IP but different ports means that it's connecting to clients that share the same IP, which is correct. But there are only 20-30 clients, and they are set to reconnect each hour (using `SocketIO.wait(seconds=60*60)` from python socketio client. It looks like when they reconnect old connection handles are not released\closed?

**Comment by miguelgrinberg** [comment]
Can you remove the reconnection logic and see if that helps?

**Comment by Nixellion** [comment]
I had to turn it on, because after a while clients were hanging. 
Here's the issue discussing this problem: https://github.com/invisibleroads/socketIO-client/issues/148

Might try another client, of course

**Comment by Nixellion** [comment]
No, all clients are in a different building, they don't share anything with the host

**Comment by Nixellion** [comment]
After 1 day and 2 hours of uptime I can confirm that the problem was in file handles. It's now at 1293, which is above the default limit of 1024. 

At this rate I estimate that it will reach current limit of 1024000 in roughly a year. 
I'm not sure, maybe I should change socketio client? Or disconnect in some other way?

I don't think that it's expected behavior to keep all connections\handles after disconnect, and use new ones after reconnect.

**Comment by miguelgrinberg** [comment]
I believe I asked this before. Do you know which handles are the ones that are constantly growing, or at least growing the fastest? Is it the file handles to your database or socket handles?

Assuming it is socket handles, are your clients connecting and disconnecting often?

**Comment by Nixellion** [comment]
I answered both of these questions before:
- Mostly Socket handles are growing, the majority of handles are sockets. sqlite database handle also grows but at a much slower rate
- Clients reconnect each hour, because otherwise python socketio clients crash with this error after a while: https://github.com/invisibleroads/socketIO-client/issues/148 Hourly reconnects are overkill, but still I can't just let them run 24\7

And again, I can think of a lot of scenarios where clients have to reconnect often, spotty connection for example.

**Comment by miguelgrinberg** [comment]
Sorry about that, you did tell me about the reconnect issue with the python-client, but I wasn't totally clear on the growth of handles between database vs sockets.

Let's assume for a moment that the leak occurs when the client reconnects. I guess what we need to find out is if the client is disconnecting gracefully or not. One possibility is that the client stops waiting on a socket but does not close it, so the socket is leaked both for the client and the server. In the client it would take a lot more time for the leak to be noticed, I would assume. Do you have a disconnect handler in your application? Does it get called at every hour for each of your clients?

**Comment by Nixellion** [comment]
Hmm... I think I see a dumb mistake I made.

```
        while not TERMINATE:
            try:
                    addr = "{}://{}".format('http', socket.gethostbyname(SERVER_IP))
                    print("Wait for connection: {}:{}".format(addr, SERVER_PORT))
                    socketIO = SocketIO(addr, SERVER_PORT)

                    miner_namespace = socketIO.define(Miner_Namespace, SOCKET_NAMESPACE)
                    # ui_namespace = socketIO.define(UI_Namespace, SOCKET_NAMESPACE)

                    socketIO.wait(seconds=RESTART_DELAY * 60)
```
I think I should either initialize it once outside, or use somethihng along the line of: 

```
with SocketIO('127.0.0.1', 8000, LoggingNamespace) as socketIO:
    socketIO.emit('bbb', {'xxx': 'yyy'}, on_bbb_response)
    socketIO.wait_for_callbacks(seconds=1)
```

**Comment by Nixellion** [comment]
It worked. I moved moved creation of `socketIO` object outside the reconnect loop. At the start of the server and initial connect of clients it had 140-141 handles, and after 12 hours it's still at 140. All clients are working, restarting `socketIO.wait()` each hour, so far no errors.

Here's client code as it is right now:

```
        if pre_flight_checks():
            print (TERMINATE)
            while not TERMINATE:
                try:
                    addr = "{}://{}".format('http', socket.gethostbyname(SERVER_IP))
                    print("Wait for connection: {}:{}".format(addr, SERVER_PORT))
                    socketIO = SocketIO(addr, SERVER_PORT)

                    client_namespace = socketIO.define(client_Namespace, SOCKET_NAMESPACE)
                    # ui_namespace = socketIO.define(UI_Namespace, SOCKET_NAMESPACE)

                    while not TERMINATE:
                        try:
                            #log.info("(Re)Starting.")
                            if pre_flight_checks():
                                socketIO.wait(seconds=RESTART_DELAY * 60)
                                print("Reconnecting... : {}:{}".format(addr, SERVER_PORT))
                                time.sleep(10)
                            else:
                                print("Update required.")

                        except:
                            print(Back.YELLOW, Fore.WHITE)
                            print("CRITICAL ERROR!")
                            traceback.print_exc()
                            log.error(traceback.format_exc())
                            print("Trying to restart socket connection.")
                            print(Back.BLACK, Fore.WHITE)
                            time.sleep(2)
                    print ("Something is not right, trying to reconnect...")
                    time.sleep(10)
                except:
                    print(Back.YELLOW, Fore.WHITE)
                    print("CRITICAL ERROR!")
                    traceback.print_exc()
                    log.error(traceback.format_exc())
                    print("Trying to restart socket connection.")
                    print(Back.BLACK, Fore.WHITE)
                    time.sleep(2)
```

(`pre_flight_checks()` checks for client version updates and that config file is in place and correct)

There's new issue, however, that at the INITIAL start clients spam this: 

```
Wait for connection: http://134.0.112.77:14864
[Connected]
Logging in...
Check version...
Update not required
Login success {'owner': 'e@mail.com'}
Login success {'owner': 'e@mail.com'}
Login success {'owner': 'e@mail.com'}5:05:03.827009
Login success {'owner': 'e@mail.com'}5:05:13.828192
Login success {'owner': 'e@mail.com'}5:05:23.876139
Login success {'owner': 'e@mail.com'}5:05:33.877163
Login success {'owner': 'e@mail.com'}5:05:43.893866
Login success {'owner': 'e@mail.com'}5:05:53.894893
Login success {'owner': 'e@mail.com'}5:06:03.911613
Login success {'owner': 'e@mail.com'}
Login success {'owner': 'e@mail.com'}5:06:13.912514
Login success {'owner': 'e@mail.com'}5:06:23.960595
Login success {'owner': 'e@mail.com'}5:07:04.011560
Login success {'owner': 'e@mail.com'}5:07:14.012782
Login success {'owner': 'e@mail.com'}
Login success {'owner': 'e@mail.com'}5:07:24.029484
Login success {'owner': 'e@mail.com'}
Login success {'owner': 'e@mail.com'}5:07:34.046140
Login success {'owner': 'e@mail.com'}
Login success {'owner': 'e@mail.com'}5:07:44.094076
Login success {'owner': 'e@mail.com'}
Login success {'owner': 'e@mail.com'}5:07:54.110732
Login success {'owner': 'e@mail.com'}5:08:04.127436
Login success {'owner': 'e@mail.com'}5:08:14.144152
```
It happens within a few seconds after initial start and then they start working as expected. But it's another issue. 

Thanks a lot for you help, time and patience! 
This issue can be closed, I suppose.

**Comment by miguelgrinberg** [comment]
Awesome, it took some time, but glad to hear the leak is contained.
